### DISCUSSION TITLE
Support for latest:^0.12

### DISCUSSION THREAD

**Discussion by merajshaikh** [warrensbox/terraform-switcher#299] [Q&A]
Hi everyone,
i have .terrform-version file which says "latest:^0.12"
which is not supported directly and I have to manually shift to the 0.12.31 version
does anyone know how can I work with this?

I can't change this to 0.12.31 as my ci/cd pipeline uses tfenv and complains about this and says it should be now formatted like this "latest:^0.12"
ERROR: Pinned terraform version needs to be in the format 'latest:^<minor version>',
       e.g. 'latest:^0.12'.

**Comment by warrensbox** [comment]
Hi @merajshaikh I am not sure if I am following. 
tfenv and tfswitch are two different tools. They use different formats.
For CICD, please refer to this: 
https://tfswitch.warrensbox.com/Continuous-Integration/

**Comment by yermulnik** [comment]
> "latest:^0.12"

By any chance what this notation means in terms of TF version? Latest in `0.12` branch (`0.12.31`) or something else?

**Reply by merajshaikh** [reply]
Hi @yermulnik  @warrensbox 
Yeah I was new to using tfenv and tfswitch on that time so i got confuesd

So tfenv has this support of latest:^0.12
which means 0.12 but the latest version which would be 0.12.31

Can we add this to tfswitch please?

**Reply by yermulnik** [reply]
Well, what you're asking for is actually to implement support for `tfenv`-specific syntax with `tfenv`-specific file (`.terraform-version`) which falls apart the HCL version constraint syntax in `latest:<regex>` part. Thus I, frankly speaking, see no sense in implementing cross-tools support to this extent (where the `latest:<regex>` syntax is specific to `tfenv` only and is not supported by Terraform as such).

I'd recommend to not try and mix different tools though and switch to using either but not both.
I also would recommend to rely upon the `required_version` from TF configuration files to allow both Terraform and 3rd-party tools like `tfswutch` and `tfenv` to align on the version of Terraform required. Going this path you can set `required_version` to `"~> 0.12"` in Terraform configuration and `tfswitch` and allegedly `tfenv` will pick that constraint up and use the latest available version of Terraform.

See https://tfswitch.warrensbox.com/Quick-Start/ for more details on usage, features and options.

**Reply by merajshaikh** [reply]
Ohh got it and I agree
thanks for the detailed info

I just have 1 more question regarding the precedence how can I give precedence to version.tf(required_version) over .terraform-version for my tfswitch install?

as to tfswitch currently  has precedenc to  .terraform-version and then rest
i can't delete or remove it as other members and ci/cd are using tfenv
it is only m1 users (me) who are using tfswitch

**Reply by yermulnik** [reply]
> I just have 1 more question regarding the precedence how can I give precedence to version.tf(required_version) over .terraform-version for my tfswitch install?

~Oh, apologies, I indeed missed precedence bit...~
The `.tfswitchrc` has precedence over `.terraform-version`, but it doesn't support version constraints 😞 
It seems like you're stuck at the moment unless your team drops the `.terraform-version` file and switches to rely upon `required_version` (from what I see `tfenv` supports that: https://github.com/tfutils/tfenv#min-required--latest-allowed).

The other option is a quirk and is somewhat that @warrensbox suggested above: create a `tfswitch` shell alias that would do something like `[ -f ".terraform-version" ] && rm ".terraform-version" && { tfswitch; git checkout --  ".terraform-version"; }` (I assumed you use Git).
